elastic/kibana
# POC - Sharing index patterns

[skip-ci]

# 🚨 Notice 2021-08-19 🚨

This POC is out of date! See #107256 instead.

-----

Note: see each commit message for more details

### Overview

This PR is a proof-of-concept to:

1. **Convert index patterns to namespaceType 'multiple-isolated'** (_"share-capable"_ objects, planned for the 8.0 release) -- this will regenerate index pattern IDs to prepare them to be shareable; it requires that the index pattern management page uses the SavedObjectsClient's `resolve` API, and display some extra UI elements in case of issues with the newly introduced Legacy URL aliases
2. **Convert index patterns to namespaceType 'multiple'** (_"shareable"_ objects, planned for some 8.x release) -- this will allow index patterns to be shared between multiple spaces; it updates the index pattern management page to display a list of spaces and the "Change shared spaces" flyout

It also includes some preliminary commits to add support for the SOC `resolve` API in the SOC abstractions that the `index_pattern_management` plugin utilizes.

-----

### Not included

This PR is intended to demonstrate the steps it will take for a plugin author to convert a saved object to become _share-capable_ and subsequently to become _shareable_. It does not include other changes such as:

* Updates to unit tests
* Updates to docs

-----

### How to test

1. Start Kibana
2. Using Dev Tools, load staging data
   <details><summary>Staging data</summary>

   ```
   POST .kibana/_bulk
   { "index": { "_id" : "index-pattern:ff959d40-b880-11e8-a6d9-e546fe2bba5f" } }
   {"index-pattern":{"fieldFormatMap":"{\"taxful_total_price\":{\"id\":\"number\",\"params\":{\"pattern\":\"$0,0.[00]\"}}}","fields":"[]","timeFieldName":"order_date","title":"kibana_sample_data_ecommerce"},"type":"index-pattern","references":[],"namespaces":["default", "bravo"],"migrationVersion":{"index-pattern":"8.0.0"},"coreMigrationVersion":"8.0.0","updated_at":"2021-03-30T13:30:10.244Z"}
   { "index": { "_id" : "index-pattern:b75d094a-3b91-5d44-848c-4ccbe1338cf7" } }
   {"index-pattern":{"fieldFormatMap":"{\"taxful_total_price\":{\"id\":\"number\",\"params\":{\"pattern\":\"$0,0.[00]\"}}}","fields":"[]","timeFieldName":"order_date","title":"kibana_sample_data_ecommerce"},"type":"index-pattern","references":[],"namespaces":["alpha"],"originId":"ff959d40-b880-11e8-a6d9-e546fe2bba5f","migrationVersion":{"index-pattern":"8.0.0"},"coreMigrationVersion":"8.0.0","updated_at":"2021-03-30T13:31:09.236Z"}
   { "index": { "_id" : "index-pattern:cc855cd9-7c1b-526c-89f8-eacad4e2b799" } }
   {"index-pattern":{"fieldFormatMap":"{\"taxful_total_price\":{\"id\":\"number\",\"params\":{\"pattern\":\"$0,0.[00]\"}}}","fields":"[]","timeFieldName":"order_date","title":"kibana_sample_data_ecommerce"},"type":"index-pattern","references":[],"namespaces":["bravo"],"originId":"ff959d40-b880-11e8-a6d9-e546fe2bba5f","migrationVersion":{"index-pattern":"8.0.0"},"coreMigrationVersion":"8.0.0","updated_at":"2021-03-30T13:31:09.476Z"}
   { "index": { "_id" : "index-pattern:87ffef05-4fd1-5f00-9751-fb28ce6d391c" } }
   {"index-pattern":{"fieldFormatMap":"{\"taxful_total_price\":{\"id\":\"number\",\"params\":{\"pattern\":\"$0,0.[00]\"}}}","fields":"[]","timeFieldName":"order_date","title":"kibana_sample_data_ecommerce"},"type":"index-pattern","references":[],"namespaces":["*"],"originId":"ff959d40-b880-11e8-a6d9-e546fe2bba5f","migrationVersion":{"index-pattern":"8.0.0"},"coreMigrationVersion":"8.0.0","updated_at":"2021-03-30T13:31:09.236Z"}
   { "index": { "_id" : "legacy-url-alias:alpha:index-pattern:ff959d40-b880-11e8-a6d9-e546fe2bba5f" } }
   {"legacy-url-alias":{"targetNamespace":"alpha","targetType":"index-pattern","targetId":"b75d094a-3b91-5d44-848c-4ccbe1338cf7"},"type":"legacy-url-alias","references":[],"migrationVersion":{},"coreMigrationVersion":"8.0.0"}
   { "index": { "_id" : "legacy-url-alias:bravo:index-pattern:ff959d40-b880-11e8-a6d9-e546fe2bba5f" } }
   {"legacy-url-alias":{"targetNamespace":"bravo","targetType":"index-pattern","targetId":"cc855cd9-7c1b-526c-89f8-eacad4e2b799"},"type":"legacy-url-alias","references":[],"migrationVersion":{},"coreMigrationVersion":"8.0.0"}
   { "index": { "_id" : "legacy-url-alias:charlie:index-pattern:ff959d40-b880-11e8-a6d9-e546fe2bba5f" } }
   {"legacy-url-alias":{"targetNamespace":"charlie","targetType":"index-pattern","targetId":"87ffef05-4fd1-5f00-9751-fb28ce6d391c"},"type":"legacy-url-alias","references":[],"migrationVersion":{},"coreMigrationVersion":"8.0.0"}
   { "index": { "_id" : "space:alpha" } }
   {"space":{"name":"Alpha","disabledFeatures":[]},"type":"space","references":[],"migrationVersion":{"space":"6.6.0"},"coreMigrationVersion":"8.0.0","updated_at":"2021-03-30T13:30:35.109Z"}
   { "index": { "_id" : "space:bravo" } }
   {"space":{"name":"Bravo","disabledFeatures":[]},"type":"space","references":[],"migrationVersion":{"space":"6.6.0"},"coreMigrationVersion":"8.0.0","updated_at":"2021-03-30T13:30:41.088Z"}
   { "index": { "_id" : "space:charlie" } }
   {"space":{"name":"Charlie","disabledFeatures":[]},"type":"space","references":[],"migrationVersion":{"space":"6.6.0"},"coreMigrationVersion":"8.0.0","updated_at":"2021-03-30T13:31:00.820Z"}
   ```
   </details>
3. Navigate to **Home**, then load "Sample eCommerce orders" data
4. Navigate to **Stack Management** > **Index Patterns** to observe two index patterns that are visible in the _Default space_
   <details><summary>Index Patterns page screenshot</summary>

   ![image](https://user-images.githubusercontent.com/5295965/113184300-01677480-9223-11eb-8fc9-40c7518c1d30.png)
   </details>

The staging data simulates the following scenario:

* With Kibana 7.12, the user created three new spaces: _Alpha_, _Bravo_, and _Charlie_
* Then, the user added eCommerce sample data to the _Default space_, then copied it to the _A/B/C spaces_ (so there are four copies of the same index pattern)
* Then, the user has bookmarked the URLs to the index pattern in each space
* Then, the user upgraded Kibana to 8.0+, which converted the index patterns and regenerated their IDs in the _A/B/C spaces_ (creating new Legacy URL aliases for any object ID that was regenerated)
* Then, the user: a. shared the index pattern from the _Default space_ to the _Bravo space_, and b. shared the index pattern from the _Charlie space_ to _All spaces_

Now you can load the following URLs to test how index pattern IDs are resolved:

1. Default space: `/app/management/kibana/indexPatterns/patterns/ff959d40-b880-11e8-a6d9-e546fe2bba5f#/?_a=(tab:indexedFields)` -- this will result in `outcome: "exactMatch"`, with no changes in the UI
2. Alpha space: `/s/alpha/app/management/kibana/indexPatterns/patterns/ff959d40-b880-11e8-a6d9-e546fe2bba5f#/?_a=(tab:indexedFields)` -- this will result in `outcome: "aliasMatch"`, and it will redirect the user to the new URL for the updated object ID (`b75d094a-3b91-5d44-848c-4ccbe1338cf7`) and display a toast message
   <details><summary>Toast screenshot</summary>

   ![image](https://user-images.githubusercontent.com/5295965/113183836-7ab29780-9222-11eb-8e3f-ea02a24a42c0.png)
   </details>
3. Bravo space: `/s/bravo/app/management/kibana/indexPatterns/patterns/ff959d40-b880-11e8-a6d9-e546fe2bba5f#/?_a=(tab:indexedFields)` -- this will result in `outcome: "conflict"`, and it will show a callout to the user at the top of the page with an appropriate warning message
   <details><summary>Callout screenshot</summary>

   ![image](https://user-images.githubusercontent.com/5295965/113184004-af265380-9222-11eb-89dd-31e29def8324.png)
   </details>

If you go back to the Index Patterns page and share an object, you can observe that the index pattern is refreshed and the "Shared spaces" column is updated accordingly.
